euphonic/util.py

```diff
@@ -298,6 +298,129 @@ def mode_gradients_to_widths(mode_gradients: Quantity, cell_vectors: Quantity
+    coordinates as r in the e^-iq.r phase (Phonopy-like), to a
+    matrix which uses the cell origin coordinates as r in the
+    e^-iq.r phase (Euphonic,CASTEP-like). Also changes the shape
+    of the matrix to be compatible with Euphonic.
+
+    Parameters
+    ----------
+    force_constants
+        Shape (n_atoms, n_cells*n_atoms, 3, 3) or
+        (n_cells*n_atoms, n_cells*n_atoms, 3, 3) float ndarray with
+        phases based on the atom coordinate. n_atoms is the number
+        of atoms in the unit cell, n_cells in the number of unit cells
+        in the supercell
+    atom_r
+        Shape (n_atoms, 3) float ndarray. The coordinates of the atoms
+        in the unit cell in fractional coordinates of the unit cell
+        vectors
+    sc_atom_r
+        Shape (n_atoms_in_sc, 3) float ndarray. The coordinates of the
+        atoms in the supercell in fractional coordinates of the unit
+        cell vectors. n_atoms_in_sc is the number of atoms in the
+        supercell
+    uc_to_sc_atom_idx
+        Shape (n_atoms,) int ndarray. For each atom in the unit cell,
+        the corresponding index in the supercell
+    sc_to_uc_atom_idx
+        Shape (n_atoms_in_sc,) int ndarray. For each atom in the
+        supercell, the corresponding index in the unit cell
+    sc_matrix
+        Shape (3, 3) int ndarray. The matrix to convert from the unit
+        cell vectors to the supercell vectors
+
+    Returns
+    -------
+    fc_converted
+        Shape (n_cells, 3*n_atoms, 3*n_atoms) float ndarray. The force
+        constants matrix with a phase and shape compatible with
+        Euphonic
+    cell_origins
+        Shape (n_cells_in_sc, 3) int ndarray. The origin coordinates of
+        each unit cell within the supercell, in units of the unit cell
+        vectors
+    """
+    n_atoms_sc = len(sc_to_uc_atom_idx)
+    n_atoms_uc = len(uc_to_sc_atom_idx)
+    n_cells = int(np.rint(np.absolute(np.linalg.det(sc_matrix))))
+    if n_atoms_sc/n_atoms_uc - n_cells != 0:
+        raise ValueError(
+            f'Inconsistent numbers of cells in the supercell, unit '
+            f'cell has {n_atoms_uc} atoms, and supercell has '
+            f'{n_atoms_sc} atoms, but sc_matrix determinant suggests '
+            f'there should be {n_cells} cells in the supercell')
+    # Get cell origins for all atoms
+    cell_origins_per_atom = sc_atom_r - atom_r[sc_to_uc_atom_idx]
+    non_int = np.where(np.abs(cell_origins_per_atom
+                              - np.rint(cell_origins_per_atom)) > 1e-5)[0]
+    if len(non_int) > 0:
+        raise RuntimeError(
+            f'Non-integer cell origins for atom(s) '
+            f'{", ".join(non_int.astype(str))}, '
+            f'check coordinates and indices are correct')
+    cell_origins_per_atom = np.rint(cell_origins_per_atom).astype(np.int32)
+    # Recenter cell origins onto atom 0
+    atom0_idx = np.where(sc_to_uc_atom_idx == 0)[0]
+    cell_origins_per_atom -= cell_origins_per_atom[[atom0_idx[0]]]
```

Comment:
Good point about `atom0_idx[0]` not necessarily being the one in cell 0, I'll update it to use `uc_to_sc_atom_idx` to figure out which is the one in cell 0, which should be more reliable.

The reason we have to do this shift in the first place is due to the `e^iq.r_a` in the interpolation, if it was changed to be `e^iq.(r_a - r_0)` then then _I think_ it could be avoided... I'm not sure where is the best place do it, do an origin shift here or update the interpolation. If the origin shift is kept, I don't think the cell positions being in the 'wrong' unit cell is an issue, as we don't actually use the supercell atom positions in any further calculations. Also, the (2, 2, 2) range are supercell vectors relative to the cell origins, so as long as all cell origins have been shifted by the same amount I think it should just work.

The CaHgO2 example has this origin shift if you're interested. I've also uploaded a script I was using to help visualise some of the Phonopy supercells to pace-developers, its a bit rough but works (with this Euphonic branch) https://github.com/pace-neutrons/pace-developers/commit/392dcd3951396207f3e177da3d14db0270506cc1